src/Features/Core/Portable/Shared/Extensions/DocumentExtensions.cs

```diff
@@ -62,31 +62,28 @@ public static Document ReplaceNode<TNode>(this Document document, SyntaxNode roo
             return document.WithSyntaxRoot(newRoot);
         }
 
-        public static async Task<IEnumerable<T>> GetUnionItemsFromDocumentAndLinkedDocumentsAsync<T>(
+        public static async Task<ImmutableArray<T>> GetUnionItemsFromDocumentAndLinkedDocumentsAsync<T>(
             this Document document,
             IEqualityComparer<T> comparer,
-            Func<Document, CancellationToken, Task<IEnumerable<T>>> getItemsWorker,
+            Func<Document, Task<ImmutableArray<T>>> getItemsWorker,
             CancellationToken cancellationToken)
         {
-            var linkedDocumentIds = document.GetLinkedDocumentIds();
-            var itemsForCurrentContext = await getItemsWorker(document, cancellationToken).ConfigureAwait(false) ?? SpecializedCollections.EmptyEnumerable<T>();
-            if (!linkedDocumentIds.Any())
-            {
-                return itemsForCurrentContext;
-            }
+            using var _ = ArrayBuilder<Task<ImmutableArray<T>>>.GetInstance(out var tasks);
+            tasks.Add(Task.Run(() => getItemsWorker(document), cancellationToken));
+
+            foreach (var linkedDocumentId in document.GetLinkedDocumentIds())
+                tasks.Add(Task.Run(() => getItemsWorker(document.Project.Solution.GetRequiredDocument(linkedDocumentId)), cancellationToken));
```

Comment:
❗ Parallelism strategy change needs to be in a standalone PR

Reply:
I think this might be less of an issue compare to the similar change @CyrusNajmabadi made in keyword completion, since the work here is mostly likely to be completed asynchronously and the number of linked document is much smaller.

Reply:
i don't mind moving into another pr.

Reply:
have pulled out the parallelism bits.